### DISCUSSION TITLE
[V5] Can cors origin accept an async function?

### DISCUSSION THREAD

**Discussion by rodrigorz** [adonisjs/core#4167] [Help] [answered]
I want my client to be able to control which domains access their API.

To do this I need to get the `ALLOWED_ORIGINS` from a table inside the Database, but the origin function cannot be async.

I think the method that handles cors origins is `computeResponseOrigin` inside `@adonisjs/core/build/src/Hooks/Cors` and can be async, also `allowedOrigins(origin, ctx)` inside `computeResponseOrigin`.

For now I'm enabling any origin for `request.url()` that start with `/external` and using a middleware to see if the hostname is registered in the database.
(I do not know if the hostname is the best option to see who the requester is, is the origin header better?)

**Comment by RomainLanz** [comment, marked as answer]
Hey @rodrigorz! 👋🏻 

No, it is not possible to make this method `async`.

One thing you can do is define a global variable and use it inside the `origin` option.
Then on the boot of the application, you can fetch the different origins and update the global variable.

**Reply by rodrigorz** [reply]
Hi @RomainLanz

How is the best way to create that global variable? And every time that the origins change i need to restart the application?

**Reply by RomainLanz** [reply]
> How is the best way to create that global variable?

It all depends on you; I would personally create some kind of memory cache using a `Map`.

> And every time that the origins change i need to restart the application?

Nope, you can recompute the value and update the variable again.

**Reply by rodrigorz** [reply]
Okay, I'll do it this way.

Thanks for the answer. 😄
